driver-core/src/main/java/com/datastax/driver/core/CodecRegistry.java

```diff
@@ -0,0 +1,760 @@
+
+    /**
+     * Register the given codec with this registry.
+     *
+     * @param codec The codec to add to the registry.
+     * @param includeDerivedCollectionCodecs If {@code true}, register not only the given codec
+     * but also collection codecs whose element types are handled by the given codec.
+     * @return this CodecRegistry (for method chaining).
+     */
+    public CodecRegistry register(TypeCodec<?> codec, boolean includeDerivedCollectionCodecs) {
+        return register(Collections.singleton(codec), includeDerivedCollectionCodecs);
+    }
+
+    /**
+     * Register the given codecs with this registry.
+     *
+     * @param codecs The codecs to add to the registry.
+     * @return this Builder (for method chaining).
+     */
+    public CodecRegistry register(TypeCodec<?>... codecs) {
+        return register(Arrays.asList(codecs), false);
+    }
+
+    /**
+     * Register the given codecs with this registry.
+     *
+     * @param codecs The codecs to add to the registry.
+     * @return this Builder (for method chaining).
+     */
+    public CodecRegistry register(Iterable<? extends TypeCodec<?>> codecs) {
+        return register(codecs, false);
+    }
+
+    /**
+     * Add the given codecs to the list of known codecs of this registry.
+     *
+     * @param codecs The codecs to add to the registry.
+     * @param includeDerivedCollectionCodecs If {@code true}, register not only the given codecs
+     * but also collection codecs whose element types are handled by the given codecs.
+     * @return this Builder (for method chaining).
+     */
+    public CodecRegistry register(Iterable<? extends TypeCodec<?>> codecs, boolean includeDerivedCollectionCodecs) {
+        for (TypeCodec<?> codec : codecs) {
+            if(includeDerivedCollectionCodecs){
+                // a map keys and values of the same given type
+                this.codecs.add(0, new MapCodec(codec, codec));
+                // map codecs for combinations of the given type with all primitive types
+                for (TypeCodec<?> primitiveCodec : PRIMITIVE_CODECS) {
+                    this.codecs.add(0, new MapCodec(primitiveCodec, codec));
+                    this.codecs.add(0, new MapCodec(codec, primitiveCodec));
+                }
+                this.codecs.add(0, new ListCodec(codec));
+                this.codecs.add(0, new SetCodec(codec));
+            }
+            this.codecs.add(0, codec);
+        }
+        return this;
+    }
+
+    /**
+     * Register the given codec with this registry as an "overriding" codec.
+     * <p>
+     * This codec will only be used if the value being serialized or deserialized
+     * belongs to the specified qualified name.
+     * <p>
+     * Qualified names shoud be specified in the form of {@code keyspace.table.column} or {@code keyspace.usertype.field}.
+     * <p>
+     * <strong>IMPORTANT</strong>: Overriding codecs can only be used in conjunction with {@link GettableData} instances,
```

Comment:
My first reaction to this is "Why does this restriction exist?".  I saw the reason why in the PR description, but  It would be good to document why (may already be done).